lib/charms/opensearch/v0/opensearch_relation_provider.py

```diff
@@ -433,3 +435,70 @@ def update_dashboards_password(self):
         pwd = self.secrets.get(Scope.APP, self.secrets.password_key(KibanaserverUser))
         for relation in self.dashboards_relations:
             self.opensearch_provides.set_credentials(relation.id, KibanaserverUser, pwd)
+
+    def _put_relation_user(
+        self, user: str, permissions: dict[str], hashed_pwd: str, relation_id: int
+    ):
+        """Create a relation user.
+
+        Relation users are registered with a dedicated role which maps to the username,
+        and their name is saved in the databag for later reference.
+        """
+        self.user_manager.create_role(role_name=user, permissions=permissions)
+        if not (users := self.charm.peers_data.get_object(Scope.APP, ClientUsersDict)):
+            users = {}
+
+        if users.get(relation_id):
+            logger.warning(
+                "User %s is already registered in Peer Relation data for relation %d.",
+                user,
+                relation_id,
+            )
+
+        self.user_manager.create_user(user, [user], hashed_pwd)
+        users[str(relation_id)] = user
+        self.charm.peers_data.put_object(Scope.APP, ClientUsersDict, users)
+
+    def remove_lingering_relation_users_and_roles(  # noqa: C901
+        self, departed_relation_id: int | None = None
+    ):
+        """Removes lingering relation users and roles from opensearch.
+
+        Args:
+            departed_relation_id: if a relation is departing, pass in the ID and its user will be
+                deleted.
+        """
+        if not self.opensearch.is_node_up() or not self.unit.is_leader():
+            return
+
+        if not (relation_users := self.charm.peers_data.get_object(Scope.APP, ClientUsersDict)):
+            relation_users = {}
```

Comment:
Similar nit as above, should help reduce 1 level of indentation

Reply:
Done.